### DISCUSSION TITLE
What is the performance of cocotb, pyuvm and pyvsc and related python solutions?

### DISCUSSION THREAD

**Discussion by davidmlw** [cocotb/cocotb#3336] [Q&A]
https://github.com/chipsalliance/riscv-dv/blob/08b12066b34c9728f706e45098ba502a36d7ca59/pygen/pygen_src/README.md?plain=1#L98-L100
I read some performance concerns. I don't know the performance level of 10k instructions per 12 minutes. But if the Python side was the performance bottleneck, it would let people hesitate.

**Comment by mballance** [comment]
@davidmlw, performance can be a concern when looking at Python independent of the application. Because Python is an interpreted dynamically-typed language, it is slower than some other languages. That said, in many cases performance is either not the most important thing, or the performance of the Python portion of a system isn't the gating factor. When using Python for functional verification in simulation, both may be true. The simulation speed of a large complex design is often such that the Python portion of the environment isn't the performance bottleneck. Many would argue that the productivity boost of writing in Python and having the ability to leverage the large ecosystem of Python packages more than makes up for any performance degradation by making (expensive) engineers more productive.
